/**
  * Protocol between core and GUI.
  * When a command is sent to the core, one or more responds will be sent back,
  * there cannot be other message like 'event message' between theses messages.
  * Version : 1
  * All string are encoded in UTF-8.
  */

import "common.proto";
import "core_settings.proto";

package Protos.GUI;

/***** Core state *****/
// GUI -> Core
// id : 0xa1
// Ask to receive a new state.
// This message doesn't have a body.

// Core -> GUI
// id : 0x01
// The state is sent periodically to the GUI (each second for example)
// or right after a GUI command.
// The size of this message should be lesser than 63kB to fit in one IP packet.
message State {
   // A download can be a file or a directory.
   message Download {
      enum Status {
         QUEUED = 0x1;
         GETTING_THE_HASHES = 0x2;
         DOWNLOADING = 0x3;
         COMPLETE = 0x4;
         PAUSED = 0x5;
         DELETED = 0x6; // Should not be seen by the user.
         
         UNKNOWN_PEER_SOURCE = 0x10; // The source peer isn't found and we don't have all the hashes.
         ENTRY_NOT_FOUND = 0x11; // The entry isn't found in the source peer shared directories.
         NO_SOURCE = 0x12; // All the hashes are known but one ore more chunk are not present among the other peers.
         
         // All the following status will pause the download.
         NO_SHARED_DIRECTORY_TO_WRITE = 0x20;
         NO_ENOUGH_FREE_SPACE = 0x21;
         UNABLE_TO_CREATE_THE_FILE = 0x22;
         UNABLE_TO_RETRIEVE_THE_HASHES = 0x23;
         
         TRANSFERT_ERROR = 0x24;
         UNABLE_TO_OPEN_THE_FILE = 0x25;
         FILE_IO_ERROR = 0x26;
         FILE_NON_EXISTENT = 0x27;
         GOT_TOO_MUCH_DATA = 0x28;
         HASH_MISSMATCH = 0x29;
      }
      required uint64 id = 1; // Cannot be 0.
      
      // When the entry is physically created, Common.Entry.exists is set to true.
      // The hashes are never sent to save a bit of bandwidth.
      required Common.Entry local_entry = 2;
      
      required Status status = 3;
      required uint64 downloaded_bytes = 4;
      
      repeated Common.Hash peer_id = 5; // The first one always corresponds to the peer source.
      optional string peer_source_nick = 6;
   }
   message Upload {
      required uint64 id = 1;
      required Common.Entry file = 2; // As the download, the hashes aren't sent to save bandwidth.
      required uint32 current_part = 3;
      required uint32 nb_part = 4;
      required uint32 progress = 5; // 0 to 10000.
      required Common.Hash peer_id = 6;
   }
   message Stats {
      enum CacheStatus {
         LOADING_CACHE_IN_PROGRESS = 0; // Initial pass: synchronize the cache with the file system.
         SCANNING_IN_PROGRESS = 1; // 1st pass: grab all file/dir names and their size and index it.
         HASHING_IN_PROGRESS = 2; // 2nd pass: compute the hashes of all files.
         UP_TO_DATE = 3;
         UNKNOWN = 4;
      }
      required CacheStatus cache_status = 1;
      required uint32 progress = 2; // 0 to 10000.
      required uint32 download_rate = 3; // [byte/s].
      required uint32 upload_rate = 4; // [byte/s].
   }
   message Peer {
      required Common.Hash peer_id = 1;
      required string nick = 2;
      required uint64 sharing_amount = 3; // [byte].
      optional Common.IP ip = 4;
      optional string core_version = 5;
   }
   message SharedDir {
      required Common.Hash id = 1;
      required string path = 2; // Always ended with a '/'.
      required uint64 size = 3;
      required uint64 free_space = 4;
   }
   
   required Peer myself = 1;
   repeated Peer peer = 2;
   
   repeated SharedDir shared_directory = 3;
   
   required bool integrity_check_enabled = 7;
   
   optional bool password_defined = 10 [default = false];
      
   repeated Download download = 4;
   repeated Upload upload = 5;
   
   optional Stats stats = 6;
   
   optional Common.Interface.Address.Protocol listenAny = 8; // If this value is set then all the interfaces of the given protocol are listened. Otherwise see the following repeated 'interface' field.
   repeated Common.Interface interface = 9;
}
// GUI -> Core
// id : 0x02
// Empty message to tell the core that we have finished to process the state message.


/***** Events *****/
// Core -> GUI
// id : 0x11
message EventChatMessages {
   message Message {
      required uint64 time = 1; // In [ms] since Epoch.
      required Common.Hash peer_id = 2;
      required string message = 3;
   }
   repeated Message message = 1;
}

// Core -> GUI
// id : 0x12
message EventLogMessage {
   enum Severity {
      SV_FATAL_ERROR = 0x01;
      SV_ERROR = 0x02;
      SV_WARNING = 0x04; // Send to the GUI but should not be shown to the user.
      SV_DEBUG = 0x08; // Never send to the GUI, there are only logged in the core log file.
      SV_END_USER = 0x10;
      SV_UNKNOWN = 0x20;
   }
   
   required uint64 time = 1; // In [ms] since Epoch.
   required Severity severity = 2;
   required string message = 3;
   
   // Only in debug mode. (not used at all for the moment).
   // optional string component = 4;
}

/***** Commands. *****/
// Authentication.
// Always send by the core when a new connection is established.
// All other messages can't be sent or received without a proper authentication.
// A new salt is randomly generated each time a new connection is created.
// Core -> GUI
// id : 0x21
message AskForAuthentication {
   required uint64 salt = 1;
   required uint64 salt_challenge = 2;
}
// Password is optional for local connections.
// GUI -> Core
// id : 0x22
message Authentication {
   optional Common.Hash password_challenge = 1; // hash(password) + old_salt + salt_challenge
}
// Core -> GUI
// id : 0x23
// If BAD_PASSWORD the core should wait a bit before sending the result (for example 1s) to avoid dictionary attack.
message AuthenticationResult {
   enum Status {
      AUTH_OK = 0x01;
      AUTH_BAD_PASSWORD = 0x02;
      AUTH_PASSWORD_NOT_DEFINED = 0x03; // The core doesn't have defined a password (empty password).
      AUTH_ERROR = 0xFF;
   }
   required Status status = 1;
}

// GUI -> Core
// id : 0xB1
// Set the language of the core.
message Language {
   required Common.Language language = 1;
}

// GUI -> Core
// id : 0xD1
// Change the core password.
// Both sent passwords are salted with 'AskForAuthentication.new_salt'.
// If 'new_password' is set to a null hash, the password is removed, in this case the old password isn't needed.
message ChangePassword {
   required Common.Hash new_password = 1; // Salted with the new salt.
   required uint64 new_salt = 2;
   
   optional Common.Hash old_password = 3; // Salted with the old salt. Required only if a password has already been set. 'See State.password_defined'.
}

// GUI -> Core
// id : 0x31
// Set the settings.
message CoreSettings {
   message SharedDirectories {
      repeated string dir = 1; // A directory always ends with a slash ('/').
   }
   
   optional string nick = 1;
   optional SharedDirectories shared_directories = 2;
   optional bool enable_integrity_check = 3;   
   
   optional string listen_address = 4 [default = ""]; // If address is empty then listen to any adresses, in this case the protocol is given by 'listenAny'.
   optional Common.Interface.Address.Protocol listen_any = 5 [default = IPv6];
}

// GUI -> Core
// id : 0x41
message Search  {
   required string pattern = 2;
}
// Core -> GUI (directly)
// id : 0x42
message Tag  {
   required uint64 tag = 2;
}
// Core -> GUI (deferred)
// This message is sent for each UDP datagram received.
// id : 0x43
// Common.FindResult

// GUI -> Core
// id : 0x51
message Browse {
   required Common.Hash peer_id = 1;
   optional Common.Entries dirs = 2;
   optional bool get_roots = 3 [default = false]; // If true the roots directories will be appended to the entries result. If the Dir field above is empty then the roots folders will always be sent whatever get_roots is true or false.
}
// Core -> GUI (directly)
// id : 0x52
// Tag
// Core -> GUI (deferred)
// id : 0x53
message BrowseResult {
   required uint64 tag = 1;
   repeated Common.Entries entries = 2; 
}

// GUI -> Core
// id : 0x61
// Remove one or several download from the queue. If this download
// is not finished the unfinished file is removed.
message CancelDownloads {
   repeated uint64 id = 1 [packed = true];
   optional bool complete= 2 [default = false];
}

// GUI -> Core
// id : 0xC1
// Pause or unpause the given downloads.
message PauseDownloads {
   repeated uint64 id = 1 [packed = true];
   optional bool pause = 2 [default = true]; // Id false -> unpause.
}

// GUI -> Core
// id : 0x71
// Tell the core to move one or more downloads in the list right before or after a given set of downloads.
// 'id_ref' and 'id_to_move' must have at least one item each.
// The order of the moved downloads are kept unchanged.
//
// Example #1: id_ref = [21], id_to_move = [16, 18], position = BEFORE
// Initial list: [34, 21, 18, 54, 16]
// After the operation: [34, 18, 16, 21, 54]
//
// Example #2: id_ref = [8, 2, 5], id_to_move = [3, 6], position = AFTER
// Initial list : [3, 8, 5, 6, 2]
// After the operation: [8, 5, 2, 3, 6]
message MoveDownloads {
   enum Position {
      BEFORE = 1;
      AFTER = 2;
   }
   repeated uint64 id_ref = 1;
   repeated uint64 id_to_move = 2;
   optional Position position = 3 [default = BEFORE];
}

// GUI -> Core
// id : 0x81
// Says to the core to download a file or a directory.
// The given path can be relative to a shared directory (if destination_directory_id is set)  or absolute (if destination_directory_id is null).
message Download {
   required Common.Hash peer_id = 1; // From where the file/dir is.
   required Common.Entry entry = 2;
   
   optional Common.Hash destination_directory_id = 3; // Where to put the folder/file. It must one of the directory of State.destination_directory. If this directory is not set, the first directory with enough free space will be taken.
   optional string destination_path = 4 [default="/"]; // Must ended with an '/'. If 'Common.Hash destination_directory_id' isn't set then a new shred directory will be created (if needed).
}

// GUI -> Core
// id : 0x91
// Send a chat message to everyone.
message ChatMessage {
   required string message = 1;
}

